### PR TITLE
feat: add kanban shortcuts

### PR DESCRIPTION
**Pull request by richardshiue**
resolves https://github.com/AppFlowy-IO/AppFlowy/issues/4995
resolves https://github.com/AppFlowy-IO/AppFlowy/issues/2856
resolves https://github.com/AppFlowy-IO/AppFlowy/issues/5283
resolves https://github.com/AppFlowy-IO/AppFlowy/issues/3249

### Feature Preview

---

#### PR Checklist

- [x] My code adheres to [AppFlowy's Conventions](https://docs.appflowy.io/docs/documentation/software-contributions/conventions)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

### CODE REVIEW

File: frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+          .add(GroupedRowId(rowId: lastCard.id, groupId: lastGroup!.id));
+    }
+  }
+
+  void _deepCopy() {
+    _focusedCards = [..._focusedCards];
+  }
+}
+
+class ConditionalListenableBuilder<T> extends StatefulWidget {
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart:372]: Let's move this to `appflowy_flutter/lib/shared/conditional_listenable_builder.dart`

File: frontend/appflowy_flutter/test/bloc_test/board_test/create_card_test.dart [hunk 2]
```diff
@@ -15,26 +15,35 @@ void main() {
...
     final databaseController = DatabaseController(view: context.gridView);
     final boardBloc = BoardBloc(
-      view: context.gridView,
       databaseController: databaseController,
     )..add(const BoardEvent.initial());
     await boardResponseFuture();
 
-    final groupId = boardBloc.state.groupIds.last;
+    List<String> groupIds = boardBloc.state.maybeMap(
+      orElse: () => [],
```
Comment (Xazin) [inline frontend/appflowy_flutter/test/bloc_test/board_test/create_card_test.dart:23]:
```suggestion
      orElse: () => const [],
```

File: frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+            rowId: iterable.elementAt(1).id,
+            groupId: lastFocusedCard.groupId,
+          ),
+        );
+    }
+
+    notifyListeners();
+  }
+
+  void focusPrevious() {
```
[the remark this answers is not in the document]
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart:100]: I use this for arrow up

image

File: frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+    _deepCopy();
+    if (_focusedCards.contains(groupedRowId)) {
+      _focusedCards.remove(groupedRowId);
+    } else {
+      _focusedCards.add(groupedRowId);
+    }
+    notifyListeners();
+  }
+
+  void focusNext() {
```
[the remark this answers is not in the document]
Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart:47]: Oh, must have missed it :pray:

File: frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart (hunk 1 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart]: Done!

File: frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+    _deepCopy();
+    _focusedCards.addAll(groupedRowIds);
+    notifyListeners();
+  }
+
+  void unFocus(List<GroupedRowId> groupedRowIds) {
+    _deepCopy();
+    _focusedCards.removeWhere(groupedRowIds.contains);
+    notifyListeners();
+  }
```
[the remark this answers is not in the document]
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart]: Done!

File: frontend/appflowy_flutter/lib/mobile/presentation/database/board/mobile_board_page.dart [hunk 6]
```diff
@@ -3,38 +3,82 @@ import 'package:appflowy/generated/locale_keys.g.dart';
...
+
+class _MobileBoardPageState extends State<MobileBoardPage> {
+  @override
+  Widget build(BuildContext context) {
+    return BlocProvider<BoardBloc>(
+      create: (_) => BoardBloc(
+        databaseController: widget.databaseController,
+      )..add(const BoardEvent.initial()),
+      child: BlocBuilder<BoardBloc, BoardState>(
+        buildWhen: (p, c) => c.isReady,
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/mobile/presentation/database/board/mobile_board_page.dart:50]: Would this be better
```suggestion
        buildWhen: (p, c) => !p.isReady && c.isReady,
```

File: frontend/appflowy_flutter/lib/mobile/presentation/database/board/mobile_board_page.dart [hunk 7]
```diff
@@ -58,50 +102,63 @@ class _MobileBoardContentState extends State<MobileBoardContent> {
...
-            ),
-            footerBuilder: _buildFooter,
-            cardBuilder: (_, column, columnItem) => _buildCard(
-              context: context,
-              afGroupData: column,
-              afGroupItem: columnItem,
-              cardMargin: config.cardMargin,
-            ),
+          return state.maybeMap(
+            orElse: () => const SizedBox.shrink(),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/mobile/presentation/database/board/mobile_board_page.dart:124]: Should this have an info widget?
```suggestion
            orElse: () => FlowyMobileStateContainer.info(
              emoji: '🛸',
              title: LocaleKeys.board_mobile_failedToLoad.tr(),
            ),
```

File: frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart [hunk 8]
```diff
@@ -150,88 +160,97 @@ class MobileHiddenGroup extends StatelessWidget {
 
     return BlocBuilder<BoardBloc, BoardState>(
       builder: (context, state) {
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart:162]: We have a few `BoardBloc` builders, should we add some logic to `buildWhen` to stop rebuilds to happen unnecessarily?

File: frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart [hunk 9]
```diff
@@ -150,88 +160,97 @@ class MobileHiddenGroup extends StatelessWidget {
...
-          (g) => g.groupId == this.group.groupId,
-        );
-        if (group == null) {
-          return const SizedBox.shrink();
-        }
+        return state.maybeMap(
+          orElse: () => const SizedBox.shrink(),
+          ready: (state) {
+            final group = state.hiddenGroups.firstWhereOrNull(
+              (g) => g.groupId == this.group.groupId,
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart:167]: Do we need to use `this` here?

File: frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart [hunk 10]
```diff
@@ -150,88 +160,97 @@ class MobileHiddenGroup extends StatelessWidget {
...
+                      content: FlowyText(
+                        LocaleKeys.board_mobile_showGroupContent.tr(),
+                      ),
+                      actionButtonTitle: LocaleKeys.button_yes.tr(),
+                      actionButtonColor: Theme.of(context).colorScheme.primary,
+                      onActionButtonPressed: () =>
+                          context.read<BoardBloc>().add(
+                                BoardEvent.toggleGroupVisibility(
+                                  group,
+                                  true,
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart:244]: Do we need to send this boolean with the event? Toggle should just reverse the existing state.

File: frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart (hunk 9 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart:167]: Done

File: frontend/appflowy_flutter/lib/mobile/presentation/database/board/mobile_board_page.dart (hunk 7 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/mobile/presentation/database/board/mobile_board_page.dart:124]: Correct me if I'm wrong but as I understand it this case isn't reachable? The parent widget should handle the other cases

File: frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart (hunk 8 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart:162]: Removed one

File: frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart (hunk 10 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/mobile/presentation/database/board/widgets/mobile_hidden_groups_column.dart:244]: I renamed this event

**Pull request merged**